react-components/src/components/Reveal3DResources/queryMappedData.ts

```diff
@@ -0,0 +1,169 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+import { type Cognite3DViewer, type PointerEventData, type CogniteCadModel } from '@cognite/reveal';
+import { type CogniteInternalId, type CogniteClient } from '@cognite/sdk';
+import {
+  type EdgeItem,
+  type InspectResultList,
+  type FdmSDK,
+  type DmsUniqueIdentifier,
+  type Source
+} from '../../utilities/FdmSDK';
+import { type FdmAssetMappingsConfig } from '../../hooks/types';
+import { type NodeDataResult } from './types';
+
+async function getAncestorNodeIdsForTreeIndex(
+  client: CogniteClient,
+  model: CogniteCadModel,
+  treeIndex: number
+): Promise<CogniteInternalId[]> {
+  const nodeId = await model.mapTreeIndexToNodeId(treeIndex);
+
+  const ancestorNodes = await client.revisions3D.list3DNodeAncestors(
+    model.modelId,
+    model.revisionId,
+    nodeId
+  );
+
+  return ancestorNodes.items.map((n) => n.id);
+}
+
+async function getMappingEdges(
+  fdmClient: FdmSDK,
+  fdmConfig: FdmAssetMappingsConfig,
+  model: CogniteCadModel,
+  ancestorIds: CogniteInternalId[]
+): Promise<{ edges: Array<EdgeItem<Record<string, any>>> }> {
+  const filter = {
+    and: [
+      {
+        equals: {
+          property: ['edge', 'endNode'],
+          value: {
+            space: fdmConfig.global3dSpace,
+            externalId: `model_3d_${model.modelId}`
+          }
+        }
+      },
+      {
+        equals: {
+          property: [
+            fdmConfig.source.space,
+            `${fdmConfig.source.externalId}/${fdmConfig.source.version}`,
+            'revisionId'
+          ],
+          value: model.revisionId
+        }
+      },
+      {
+        in: {
+          property: [
+            fdmConfig.source.space,
+            `${fdmConfig.source.externalId}/${fdmConfig.source.version}`,
+            'revisionNodeId'
+          ],
+          values: ancestorIds
+        }
+      }
+    ]
+  };
+
+  return await fdmClient.filterAllInstances(filter, 'edge', fdmConfig.source);
+}
+
+async function inspectNode(
+  fdmClient: FdmSDK,
+  dataNode: DmsUniqueIdentifier
+): Promise<InspectResultList> {
+  const inspectionResult = await fdmClient.inspectInstances({
+    inspectionOperations: { involvedViewsAndContainers: {} },
+    items: [
+      {
+        instanceType: 'node',
+        externalId: dataNode.externalId,
+        space: dataNode.space
+      }
+    ]
+  });
+
+  return inspectionResult;
+}
+
+async function filterNodeData<NodeType>(
+  fdmClient: FdmSDK,
+  dataNode: DmsUniqueIdentifier,
+  dataView: Source
+): Promise<NodeType | undefined> {
+  if (dataView === undefined) {
+    return undefined;
+  }
+
+  const dataQueryResult = await fdmClient.filterAllInstances(
+    {
+      and: [
+        { equals: { property: ['node', 'space'], value: dataNode.space } },
+        {
+          equals: {
+            property: ['node', 'externalId'],
+            value: dataNode.externalId
+          }
+        }
+      ]
+    },
+    'node',
+    dataView
+  );
```

Comment:
Isn't it a retrieval of the node by `externalId`? Shouldn't we use [this api call](https://api-docs.cognite.com/20230101/tag/Instances/operation/byExternalIdsInstances/) for that?

Reply:
Ah, you're right - I've just taken some shortcuts since we already had that query available